lib/wibox/widget/calendar.lua

```diff
@@ -0,0 +1,401 @@
+    return props.fn_embed(out_layout, "year")
+end
+
+
+--- Set the container to the current date
+-- @param self Widget to update
+local function fill_container(self)
+    local date = self._private.date
+    if date then
+        -- Create calendar grid
+        if not self._private.calendar_cache[date] then
+            if self.type == "month" then
+                self._private.calendar_cache[date] = create_month(self._private, date)
+            elseif self.type == "year" then
+                self._private.calendar_cache[date] = create_year(self._private, date)
+            end
+        end
+        -- Create calendar box
+        self._private.container = self._private.calendar_cache[date]
+    else
+        self._private.container = nil
+    end
+    self:emit_signal("widget::layout_changed")
+end
+
+--- Clear the cache and redraw
+-- @param self Widget to update
+local function clear_cache(self)
+    self._private.calendar_cache = {}
+    fill_container(self)
+end
+
+-- Set the calendar date
+function calendar:set_date(date)
+    if date ~= self._private.date then
+        self._private.date = date
+        -- Create calendar grid
+        fill_container(self)
+    end
+end
+
+
+-- Build properties function
+for _, prop in ipairs(properties) do
+    -- setter
+    if not calendar["set_" .. prop] then
+        calendar["set_" .. prop] = function(self, value)
+            if (string.sub(prop,1,3)=="fn_" and type(value) == "function") or self._private[prop] ~= value then
+                self._private[prop] = value
+                -- empty cache
+                clear_cache(self)
+                --self:emit_signal("widget::layout_changed")
+            end
+        end
+    end
+    -- getter
+    if not calendar["get_" .. prop] then
+        calendar["get_" .. prop] = function(self)
+            return self._private[prop]
+        end
+    end
+end
+
+-- Layout widget
+function calendar:layout(_, width, height)
+    if self._private.container then
+        return { base.place_widget_at(self._private.container, 0, 0, width, height) }
+    end
+    return {}
+end
+
+-- Fit widget
+function calendar:fit(context, width, height)
+    if self._private.container then
+        return base.fit_widget(self, context, self._private.container, width, height)
+    end
+    return 0,0
+end
+
+
+--- Return a new calendar widget by type.
+--
+-- @tparam string t Type of the calendar, `year` or `month`
+-- @tparam string date Date of the calendar
+-- @tparam[opt="Monospace 10"] string font Font of the calendar
+-- @treturn widget The calendar widget
+local function get_calendar(t, date, font)
+    if t ~= "month" and t ~= "year" then return end
+    local ret = base.make_widget(nil, "calendar_"..t, {enable_properties = true})
+    gtable.crush(ret, calendar, true)
+
+    ret.type = t
+    ret._private.calendar_cache = {}
+
+    -- default values
+    ret._private.date = date
+    ret._private.font = font or beautiful.calendar_font or "Monospace 10"
```

Comment:
i would add just `beautiful.font` before Monospace but i think monospace is good one, because if not unset it could use sans and make calendar unreadable